lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -0,0 +1,365 @@
+package invoicesrpc
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/lightningnetwork/lnd/htlcswitch"
+	"github.com/lightningnetwork/lnd/invoices"
+	"github.com/lightningnetwork/lnd/netann"
+	"github.com/lightningnetwork/lnd/route"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcutil"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/zpay32"
+)
+
+// AddInvoiceConfig contains dependencies for invoice creation.
+type AddInvoiceConfig struct {
+	// InvoiceRegistry is a central registry of all the outstanding invoices
+	// created by the daemon.
+	InvoiceRegistry *invoices.InvoiceRegistry
+
+	// Switch is used by the invoices subserver to control acceptance and
+	// cancelation of invoices.
+	Switch *htlcswitch.Switch
```

Comment:
We only need `HasActiveLink` here not the _entire_ switch. Therefore we can get by with just defining a function closure for use here, or an interface that lets us know if a link is active or not.

Reply:
Similar comments apply to the usage of `ChanDB` below.

Reply:
Do you want to do this in this PR? The dependency became more explicit because of the code move, but that is also all that happened. A move. I'd like to try to prevent this PR from expanding even further. Adding the interface can always be done later. It has no design impact at the moment imo.